src/ATen/native/transformers/xpu/flash_attn/sycltla/mha_fwd.cpp

```diff
@@ -0,0 +1,563 @@
+          EpilogueDispatchPolicy,
+          MMAOperation,
+          TileShapeOutPut,
+          SubgroupLayout,
+          ElementComputeEpilogue,
+          ElementOutput,
+          cutlass::gemm::TagToStrideC_t<LayoutO>,
+          ElementOutput,
+          GmemTiledCopyStore>;
+  using CollectiveSoftmaxEpilogue =
+      cutlass::flash_attention::collective::FlashPrefillSoftmaxEpilogue<
+          IS_CAUSAL,
+          EpilogueDispatchPolicy,
+          ElementAccumulator>;
+
+  using namespace cutlass::fmha::collective;
+
+  using ProblemShapeType = ProblemShape;
+
+  // Mainloop
+  using CollectiveMainloop =
+      cutlass::flash_attention::collective::FlashPrefillMma<
+          GEMMDispatchPolicy,
+          ProblemShapeType,
+          ElementInputQ,
+          cutlass::gemm::TagToStrideA_t<LayoutQ>,
+          ElementInputKV,
+          cutlass::gemm::TagToStrideB_t<LayoutK>,
+          ElementInputKV,
+          cutlass::gemm::TagToStrideB_t<LayoutV>,
+          MMAOperation,
+          TileShapeQK,
+          TileShapePV,
+          SubgroupLayout,
+          GmemTiledCopyQ, // Q
+          GmemTiledCopyK, // K
+          GmemTiledCopyV, // V,
+          IS_CAUSAL>;
+  using FMHAPrefillKernel = cutlass::flash_attention::kernel::FMHAPrefill<
+      ProblemShapeType,
+      CollectiveMainloop,
+      CollectiveSoftmaxEpilogue,
+      CollectiveEpilogue,
+      cutlass::flash_attention::IndividualScheduler>;
+
+  FA2Runner<FMHAPrefillKernel> runner;
+  runner.run(
+      queue, problem_shape, hw_info, query, key, value, out, logsumexp, scale);
+}
+
+template <typename T, typename ProblemShape, bool IS_CAUSAL>
+void run_mha_fwd_(
+    sycl::queue& queue,
+    ProblemShape& problem_shape,
+    const T* query,
+    const T* key,
+    const T* value,
+    T* out,
+    float* logsumexp,
+    float scale) {
+  const int headdim = get<5>(problem_shape);
+
+#define run_mha_fwd_specialized( \
+    TileShapeQK_,                \
+    TileShapePV_,                \
+    TileShapeOutPut_,            \
+    SubgroupLayout_,             \
+    PipelineStages_)             \
+  run_mha_fwd_<                  \
+      T,                         \
+      ProblemShape,              \
+      IS_CAUSAL,                 \
+      TileShapeQK_,              \
+      TileShapePV_,              \
+      TileShapeOutPut_,          \
+      SubgroupLayout_,           \
+      PipelineStages_>(          \
+      queue, problem_shape, query, key, value, out, logsumexp, scale);
+
+  constexpr int PipelineStages = 2;
+  if (headdim == 64) {
+    using TileShapeQK = Shape<_128, _64, _64>;
+    using TileShapePV = Shape<_128, _32, _64>;
+    using TileShapeOutPut = Shape<_128, _64, _64>;
+    using SubgroupLayout = Layout<Shape<_8, _1, _1>, Stride<_1, _1, _1>>;
+    run_mha_fwd_specialized(
+        TileShapeQK,
+        TileShapePV,
+        TileShapeOutPut,
+        SubgroupLayout,
+        PipelineStages);
```

Comment:
Is this template-specifialized functiom dedicated to BMG? If so, how can users pass Arch information to this function?